src/submodule.c

```diff
@@ -832,6 +832,47 @@ int git_submodule_repo_init(
 	return error;
 }
 
+static int clone_return_origin(git_remote **out, git_repository *repo, const char *name, const char *url, void *payload)
+{
+	GIT_UNUSED(url);
+	GIT_UNUSED(payload);
+	return git_remote_lookup(out, repo, name);
+}
+
+static int clone_return_repo(git_repository **out, const char *path, int bare, void *payload)
+{
+	git_submodule *sm = payload;
+
+	GIT_UNUSED(path);
+	GIT_UNUSED(bare);
+	return git_submodule_open(out, sm);
+}
+
+int git_submodule_add_clone(git_repository **out, git_submodule *submodule)
+{
+	int error;
+	git_repository *clone;
+	git_clone_options opts = GIT_CLONE_OPTIONS_INIT;
+
+	assert(submodule);
+
+	opts.repository_cb = clone_return_repo;
+	opts.repository_cb_payload = submodule;
+	opts.remote_cb = clone_return_origin;
+	opts.remote_cb_payload = submodule;
+
+	error = git_clone(&clone, git_submodule_url(submodule), git_submodule_path(submodule), &opts);
```

Comment:
I'm surprised this actually works in the test suite. The problem is that when I've just initialized a submodule via `git_submodule_add_setup`, it will already contain the ".git" gitlink to the actual gitdir of the submodule. Because of that, `git_clone` will refuse to clone into an non-empty directory. And in fact it does fail when changing my example code to use this function instead of implementing the clone step myself. Puzzled

Reply:
Never mind, now it works. I think the examples just didn't build. And in fact the `clone_return_repo` callback should be the reason why it succeeds

Reply:
Argh. I've managed to hard-reset my adjustment to use your new API, so never mind my never mind comment. It does fail in `git_clone`, line 406, where it checks whether the local path is an empty directory or not.

Reply:
Ah, I got it. `git_submodule_path` returns a relative path, which is why `git_clone` thinks that the target path doesn't yet exist. You have to create an absolute path for submodule relative to the parent repo's working directory, which will subsequently break your test as `git_clone` will error out due to files existing in the target directory.

Sorry for my confused comments